File: Leetcode/WordSearch.cpp
class Solution {
public:

    bool dfs(vector<vector<char>>& b, vector<vector<bool>>& used, string &w, int num, int i, int j, int n, int m){
        bool Ok = false;
        if(num == w.size()) return 1;
        if(!Ok && i - 1 >= 0 && !used[i - 1][j] && b[i - 1][j] == w[num]){
            used[i - 1][j] = 1;
            Ok |= dfs(b, used, w, num + 1, i - 1, j, n, m);
            used[i - 1][j] = 0;
        }
        if(!Ok && i + 1 < n && !used[i + 1][j] && b[i + 1][j] == w[num]){
            used[i + 1][j] = 1;
            Ok |= dfs(b, used, w, num + 1, i + 1, j, n, m);
            used[i + 1][j] = 0;
        }
        if(!Ok && j - 1 >= 0 && !used[i][j - 1] && b[i][j - 1] == w[num]){
            used[i][j - 1] = 1;
            Ok |= dfs(b, used, w, num + 1, i, j - 1, n, m);
            used[i][j - 1] = 0;
        }
        if(!Ok && j + 1 < m && !used[i][j + 1] && b[i][j + 1] == w[num]){
            used[i][j + 1] = 1;
            Ok |= dfs(b, used, w, num + 1, i, j + 1, n, m);
            used[i][j + 1] = 0;
        }
        return Ok;
    }

    bool exist(vector<vector<char>>& b, string w) {
        if(w.size() == 0) return 1;
        int n = b.size();
        if(n == 0) return 0;
        int m = b[0].size();
        vector<vector<bool>> used;
        used.resize(n + 1);
        for(int i = 0; i < n; ++i)used[i].assign(m + 1, 0);
        bool Ok = false;
        for(int i = 0; i < n && !Ok; ++i){
            for(int j = 0; j < m && !Ok; ++j){
                if(b[i][j] == w[0]) {
                    used[i][j] = 1;
                    Ok |= dfs(b, used, w, 1, i, j, n, m);
                    used[i][j] = 0;
                }
            }
        }
        return Ok;
    }
};

File: Leetcode/WordSearch.rs
// https://leetcode.com/problems/word-search

struct Solution {}

impl Solution {
    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        fn dfs(
            board: &Vec<Vec<char>>,
            word: &Vec<char>,
            (i, j): (usize, usize),
            pos: usize,
            used: &mut Vec<Vec<bool>>,
        ) -> bool {
            if pos == word.len() {
                return true;
            }

            used[i][j] = true;
            for (to_i, to_j) in
                [(0, 1), (0, -1), (1, 0), (-1, 0)]
                    .into_iter()
                    .filter_map(|(di, dj)| {
                        let to_i = (i as i32 + di) as usize;
                        let to_j = (j as i32 + dj) as usize;
                        let to_c = board.get(to_i)?.get(to_j)?;
                        (to_c == &word[pos]).then_some((to_i, to_j))
                    })
            {
                if !used[to_i][to_j] && dfs(board, word, (to_i, to_j), pos + 1, used) {
                    return true;
                }
            }

            used[i][j] = false;

            false
        }

        let word = word.chars().collect::<Vec<_>>();

        let mut used = vec![vec![false; board[0].len()]; board.len()];

        for (i, r) in board.iter().enumerate() {
            for (j, c) in r.iter().enumerate() {
                if c == &word[0] {
                    if dfs(&board, &word, (i, j), 1, &mut used) {
                        return true;
                    }
                }
            }
        }

        false
    }
}
